src/ln/invoice/mod.rs

```diff
@@ -0,0 +1,187 @@
+use std::str::FromStr;
+use std::num::ParseIntError;
+
+use bech32;
+use bech32::Bech32;
+
+use chrono::{DateTime, Utc, Duration};
+
+use regex::Regex;
+
+use secp256k1;
+use secp256k1::key::PublicKey;
+use secp256k1::{Signature, Secp256k1};
+
+mod parsers;
+
+/// An Invoice for a payment on the lightning network as defined in
+/// [BOLT #11](https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#examples).
+
+#[derive(Eq, PartialEq, Debug)]
+pub struct Invoice {
+    /// The currency deferred from the 3rd and 4th character of the bech32 transaction
+    pub currency: Currency,
+
+    /// The amount to pay in pico-satoshis
+    pub amount: Option<u64>,
+
+    pub timestamp: DateTime<Utc>,
+
+    /// tagged fields of the payment request
+    pub tagged: Vec<TaggedField>,
+
+    pub signature: Signature,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Currency {
+    Bitcoin,
+    BitcoinTestnet,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum TaggedField {
+    PaymentHash([u8; 32]),
+    Description(String),
+    PayeePubKey(PublicKey),
+    DescriptionHash([u8; 32]),
+    ExpiryTime(Duration),
+    MinFinalCltvExpiry(u64),
+    Fallback(Fallback),
+    Route {
+        pubkey: PublicKey,
+        short_channel_id: u64,
+        fee_base_msat: i32,
+        fee_proportional_millionths: i32,
+        cltv_expiry_delta: u16,
+    }
+}
+
+// TODO: better types instead onf byte arrays
```

Comment:
Just use rust-bitcoin's Script object directly?

Reply:
Wouldn't that have quite bad ergonomics? It forces the user to build a complete script and our library checks if it is one of the three supported types, strips away all the boilerplate/OP_CODEs and extracts the hash that the user probably had to begin with. The other way around seems to be a bit cumbersome too: how do I generate the corresponding address from the script (even worse: a `to_address()` function will have to return a `Result` because afaik not every instance of `Script` is encodable as some type of address)?

That's just how I understood `Script` from the docs, maybe I missed something.

Reply:
Ah, oops, sorry, I was just thinking the deserialize -> internal struct direction, not struct -> serialized form direction. Probably time to clean up the rust-bitcoin address stuff so that it supports more types of addresses... @apoelstra ?